### PR TITLE
orientation module shows toggled transformations

### LINKED ISSUE #12176: Provide a way to get/set image rotation and flip via the Lua API
**Is your feature request related to a problem? Please describe.**
To my knowledge, there is currently no way to rotate images from the Lua API, or to find out what rotation has been applied to an image previously.

**Describe the solution you'd like**
Add one or more functions to dt_lua_image_t to get/set the rotation of an image.

**Alternatives**
It might be possible to enable 90 degree rotation by adding an orientation attribute to dt_lua_image_t (akin to the EXIF orientation flag) so that the image automatically rotates according to the value of that attribute. This would have the advantage of mapping cleanly to the standard EXIF orientation flag. But from what I understand (per the discussion in #9912) this would require modifying other components of darktable to respect the orientation attribute.

**Additional context**
This was discussed some in #9912, which requested an orientation attribute.

### PR DESCRIPTION
**Pull request by dterrahe**
Turn rotate and flip buttons in the orientation module into toggle buttons that show whether the transformation is active.
image

This also means that midi keys mapped to these buttons light up when pressed and that for example
`dt.gui.action("iop/flip/rotate 90 degrees CW", 0, "", "toggle", "")
`
will return the current status. 

So this possibly also fixes #12176.

Holding ctrl while pressing a button will perform a reset first before applying the transformation.

### CODE REVIEW

**Comment by TurboGit** [comment]
Maybe something I'm missing, but to me the transformation are always active/possible, no?

**Comment by dterrahe** [comment]
> Maybe something I'm missing, but to me the transformation are always active/possible, no?

You may want to try it out; pushing a button is worth a 1000 words (especially when they are as vague as mine)...

**Comment by TurboGit** [comment]
Maybe just me but I found this a bit confusing. The highlight buttons looks more as if those are actives compared to the others. I would prefer probably a small underline to convey the fact that the actual transformation has been applied.

Let's see what others have to say about this! @Nilvus : Again you probably want to give your feedback on this. TIA.

**Comment by TurboGit** [comment]
Just let me stress that I like the idea, I have tagged this for 4.2, just the UI is confusing to me.

**Comment by dterrahe** [comment]
> The highlight buttons looks more as if those are actives compared to the others.

As compared to which others? This uses bog standard dtgtk_togglebuttons so should look _exactly_ the same when toggled on or off as everything else for maximum consistency. The _only_ thing that's different is that these buttons are wider (together they "fill" the available space). That can be changed by leaving "fill" off, but that creates gaps between the buttons, which to me is ugly. They could also all be bunched together (against the right border, which is the default for most buttons in other modules) by leaving "expand" off as well (basically remove the _for_ loop at the end of _gui_init_ and add the buttons in reverse order). I don't think it looks as nice and making the buttons bigger makes them easier to hit, but if you prefer, I'm certainly not religious about it. At least to me it would be _much_ better than inventing a whole new special indicator here (like underline or whatever else we could come up with).

**Comment by TurboGit** [comment]
Compared to the other buttons in the screenshot you posted. For me it looks like the button on the 1st button on the left and the last button on the right are inactive. But maybe that's just me, at least this was the confusion on my previous comment.

**Comment by dterrahe** [comment]
> But maybe that's just me

I don't know, but if you look at any other row of toggle buttons, say for overexposed or softproofing, you'll have the same situation where some are enabled and others next to them are not, but they are still "active". I'm really somewhat surprised that this would cause confusion, so I can't predict how others would react. But even if at first it is "different", hovering over the other buttons should make clear they can be pressed and I wonder if it takes more than seconds to get used to (i.e. if it is worth spending effort to "solve" the "problem").

But maybe others can chime in...

**Comment by TurboGit** [comment]
But those are not toggle buttons, they are action buttons always doing the same action each time they are pressed.

Look at those screenshot, starting from that (it is a portrait image):

image

Then clicking multiple times on the first button which is actually highlighted:

image

image

image

image

Having all those buttons changing background each time you click on one is... quite confusing if you ask me :)

**Comment by dterrahe** [comment]
> quite confusing if you ask me

Maybe :-) But what did you expect? Pressing turn ccw twice doesn't undo the action, it actually has the same effect as flipping horizontally and then vertically. Pressing it 3 times is the same as pressing cw once.

Some things can't be made simpler than they are. But if you have concrete suggestions to make this better, I'm all ears. However if it involves something that is uniquely used _just_ for this situation, I'm not sure it will.

> But those are not toggle buttons, they are action buttons always doing the same action each time they are pressed.

I _think_ what you mean to say is that the expectation for a _toggle_ button is that pressing it twice gets you back to where you were. That works for flip (hor&ver), but not for rotate (you need to press it 4 times). So maybe you feel that in that case we are not allowed to use the button to also indicate a state.

I'm not sure. I still feel that the highlighted buttons clearly show how you could have gotten to the current state by pressing just these buttons. For me it is one of those "oh yeah, hadn't thought of that but makes sense; cute!" reactions. And as soon as you press a rotate again you can immediately see from the image that indeed you turned in the same direction twice, so there's not much risk of anybody receiving long term damage. We could turn the rotates into toggles as well, but that would force the user to toggle off the 90 degree turn and turn both flips on, which would just be stupid. Or have a separate indicator or text (in the code we have TRANSPOSE and TRANSVERSE but I don't think that means much to most users nor gives an indication how to achieve it).

But if in your opinion it is really _wrong_ ... then I would ask again for your better suggestion. I can't at the moment come up with a solution I would like better, or just as much, as this, but I would love to be surprised!

**Comment by TurboGit** [comment]
> Maybe :-) But what did you expect?

Don't know :) But this doesn't look good to me. I'll try to think about other options when I have time. Maybe others will have more ideas...

**Comment by Nilvus** [comment]
I check some other software (on Linux) and none do that. All works like darktable actually.
Toggle button is for "on/off" actions, not for buttons with multiple actions possible.
I also check (with some youtube videos), how main softwares like Lightroom or Capture One works. All software I see never let such buttons active.

I fear that this would be another way to do something simple different than other software and so add confusion for users going from other software or using other ones). On all software, there's no indication about using it or not. And I don't understand why it's needed. Using that is only when you want to rotate or flip and you know (I hope) if you need or not. And if you already use it, there's history to check for it.

Thus, the issue you linked asked for a way on Lua API, not an the module.

**To summarize this:** as all other software works the way darkable actually does on master, it's a bad idea to change that.

**Comment by dterrahe** [comment]
> On all software, there's no indication about using it or not.

Indeed, and it has always annoyed me a little. But I guess we can never question the superior choices of other software.

> And if you already use it, there's history to check for it.

So if I don't remember if I flipped an image or not, it is more convenient to go to history and make sure I check if there is more than one entry for orientation, than just having it clearly shown in the module? Sure.

(I know I can just reset the module and do it again, but I thought we were worried about clueless users, who likely haven't spotted the reset button yet.)

> not for buttons with multiple actions possible.

Your argument only applies to the rotation buttons, not the flip horizontal and vertical ones. Right?

> and so add confusion for users going from other software or using other ones.

So you really think people will give up on dt because they can't figure out how to rotate an image because a rotate button is lit up, (because the image has been rotated!) after they learned all about scene referred and filmic? Maybe I have a slightly more hopeful opinion of humanity. Albeit a fading one.

**Comment by parafin** [comment]
If button has on/off state, it should be toggled by clicking on it, not by some complex logic. If you want indication of current settings - add it separately from these buttons, as a label or an icon.
P.S.
Not every idea is a good one, even if you like it very much. One should always be prepared to get negative feedback, unless you ask for none.

**Comment by dterrahe** [comment]
> One should always be prepared to get negative feedback, unless you ask for none.

I didn't actually _ask_ for feedback in this case, because to me it was such an obviously good idea! ;-) It works well with midi buttons, provides lua with a useful return value, gives visual feedback in the module so you don't accidentally leave an image flipped when you didn't mean to and in a cute way demonstrates that different sequences of transformations are equivalent.

Hearing that it _is_ confusing, while unwelcome, is useful. Being told several different ways that we have to do everything the same way others do, less so. While completely ignoring that I actually _am_ trying to fix something that I perceive as a shortcoming and not offering a better alternative...

**Comment by Nilvus** [comment]
> Hearing that it _is_ confusing, while unwelcome, is useful. Being told several different ways that we have to do everything the same way others do, less so. While completely ignoring that I actually _am_ trying to fix something that I perceive as a shortcoming and not offering a better alternative...

It's not what I mean. My main purpose was that toggle are not done for buttons with multiple states (and yes, it's especially for rotate buttons). That could make sense to use toggle for flip buttons (even if other software do not do that).
I just pointed other software because habits are strong for people and we have had complaints about other parts that darktable doesn't do the same. Anyway, I'm not to completely follow others. We can find a better way but it's harder on common things (because of habits) and here, better to do for rotate buttons at least as others.
I do not propose a better alternative because I don't have one and I don't find it useful (it's just my view of course). So maybe an alternative would be to only do what you propose here on flip buttons.

I understand and find it great that you try to fix something that you find bad. It's just that we need to take care of all views and find what could be the best for most people (and it's always hard to please all users, even ourself).

**Comment by dterrahe** [comment]
Understood and I appreciate your follow up. The "toggle buttons should toggle" rationale makes sense.

I don't think there's much interest in a flip-button-only version of this PR and I myself would rather keep using this as-is locally and not spend more time on it.
